src/NHibernate.Test/Async/NHSpecificTest/NH2319/Fixture.cs

```diff
@@ -68,12 +77,41 @@ public abstract class FixtureBaseAsync : TestCaseMappingByCode
 		}
 
 		[Test]
+		public async Task ShouldBeAbleToReuseQueryPlanAsync()
+		{
+			await (ShouldBeAbleToFindChildrenByNameAsync());
+			using (var spy = new LogSpy(typeof(QueryPlanCache)))
+			{
+				Assert.That(ShouldBeAbleToFindChildrenByNameAsync, Throws.Nothing);
+				AssertPlanCacheHit(spy);
+			}
+		}
+
+		[Test]
+		public async Task ShouldNotMixResultsAsync()
+		{
+			await (FindChildrenByNameAsync(_parent1Id, _child1Id));
+			using (var spy = new LogSpy(typeof(QueryPlanCache)))
+			{
+				await (FindChildrenByNameAsync(_parent2Id, _child3Id));
+				AssertPlanCacheHit(spy);
+			}
+		}
+
+		private static void AssertPlanCacheHit(LogSpy spy) =>
+			// Each query currently ask the cache two times, so asserting reuse requires to check cache has not been missed
+			// rather than only asserting it has been hit.
+			Assert.That(spy.GetWholeLog(),
+				Contains.Substring("located collection-filter query plan in cache (")
+				.And.Not.Contains("unable to locate collection-filter query plan in cache"));
```

Comment:
The "not contains" is not much robust: if the log message in cache is changed, this will defeat this assert. So having a cache miss test would be nice, but this has another trouble: how to be sure no other test has put an adequate plan in cache?